### PR TITLE
Align decomposition timestamps with provided `df`

### LINKED ISSUE #1202: Align decomposition timestamps with provided `df`
### 🚀 Feature Request

Align decomposition timestamps with provided `df` for:

1. `predict_components` and `forecast_components` in `_HoltWintersAdapter`
2.  `predict_components` and `forecast_components` in `_SARIMAXBaseAdapter`
3. `forecast_components` in `_TBATSAdapter`

### Proposal

Align decomposition `timestamp` with corresponding `timestamp` from provided `df`.
[This](https://github.com/tinkoff-ai/etna/blob/master/etna/models/tbats.py#L136) implementation could be used as a reference.

### Test cases

1. Test in-sample decomposition for part of training data
2. Test out-of-sample decomposition with gaps in horizon timiline
3. Test that all models with avaliable decomposition return target components

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1202

### CODE REVIEW

File: etna/models/holt_winters.py [hunk 1]
```diff
@@ -316,19 +332,24 @@ def forecast_components(self, df: pd.DataFrame) -> pd.DataFrame:
         model = self._model
         fit_result = self._result
 
-        if fit_result is None or model is None:
+        if fit_result is None or model is None or self._train_freq is None:
             raise ValueError("This model is not fitted!")
 
+        if df["timestamp"].max() <= self._last_train_timestamp:
+            raise NotImplementedError("In-sample prediction decomposition isn't supported by current implementation.")
```
Comment (alex-hse-repository) [inline etna/models/holt_winters.py:339]: Sounds misleading, we actually support in-sample decomposition, but with another method, may you specify this point in the error(everywhere, not only here)

File: etna/models/holt_winters.py [hunk 2]
```diff
@@ -207,6 +213,12 @@ def fit(self, df: pd.DataFrame, regressors: List[str]) -> "_HoltWintersAdapter":
         :
             Fitted model
         """
+        freq = pd.infer_freq(df["timestamp"], warn=False)
```
Comment (alex-hse-repository) [inline etna/models/holt_winters.py:216]: May be we cat create a separate method for this, if it is used in several models

File: etna/models/holt_winters.py [hunk 3]
```diff
@@ -316,19 +332,24 @@ def forecast_components(self, df: pd.DataFrame) -> pd.DataFrame:
         model = self._model
         fit_result = self._result
 
-        if fit_result is None or model is None:
+        if fit_result is None or model is None or self._train_freq is None:
             raise ValueError("This model is not fitted!")
 
+        if df["timestamp"].max() <= self._last_train_timestamp:
```
Comment (alex-hse-repository) [inline etna/models/holt_winters.py:338]: What if `df["timestamp"].min() <= self._last_train_timestamp`? Shouldn't all the timestamps be out of sample?

File: etna/models/holt_winters.py [hunk 4]
```diff
@@ -354,6 +375,14 @@ def forecast_components(self, df: pd.DataFrame) -> pd.DataFrame:
         if model._use_boxcox:
             components_df = self._rescale_components(components=components_df)
 
+        components_df = select_observations(
```
Comment (alex-hse-repository) [inline etna/models/holt_winters.py:378]: Don't we need to specify `start` here?

File: etna/models/utils.py [hunk 5]
```diff
@@ -50,3 +53,48 @@ def determine_num_steps(start_timestamp: pd.Timestamp, end_timestamp: pd.Timesta
...
+    start: Optional[Union[pd.Timestamp, str]] = None,
+    end: Optional[Union[pd.Timestamp, str]] = None,
+    periods: Optional[int] = None,
+) -> pd.DataFrame:
+    """Select observations from dataframe with known timeline.
+
+    Parameters
+    ----------
+    df:
+        dataframe with known timeline. If `timestamps` column is not presented `freq`, `start`,
```
Comment (alex-hse-repository) [inline etna/models/utils.py:71]: When can it happens that there is not timestamp column in df?

File: etna/models/tbats.py [hunk 6]
```diff
@@ -131,6 +132,10 @@ def forecast_components(self, df: pd.DataFrame) -> pd.DataFrame:
         raw_components = self._decompose_forecast(horizon=horizon)
```
Comment (alex-hse-repository) [inline etna/models/tbats.py:132]: What about checking the timestamps in this method?

File: tests/test_models/test_sarimax_model.py [hunk 7]
```diff
@@ -229,10 +231,55 @@ def test_components_sum_up_to_target(
...
+
+    pred_df = train.iloc[5:-5]
+
+    components = model.predict_components(df=pred_df)
+    pred = model.predict(df=pred_df, prediction_interval=False, quantiles=[])
+
+    np.testing.assert_allclose(np.sum(components.values, axis=1), np.squeeze(pred))
+
+
+def test_forecast_components_of_subset_error(dfs_w_exog):
```
Comment (alex-hse-repository) [inline tests/test_models/test_sarimax_model.py:258]: Shouldn't it work in this case?  I mean if we have long enough regressors?

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: tests/test_models/test_sarimax_model.py (hunk 7 above)
Comment (brsnw250) [reply tests/test_models/test_sarimax_model.py:258]: In the case of SARIMAX forecast decomposition, we need to estimate full horizon. Here we cut one timestamp from the beginning and providing 3 points, while 4 expected.

File: etna/models/holt_winters.py (hunk 1 above)
Comment (brsnw250) [reply etna/models/holt_winters.py]: Sure, it should be `min` to cover all cases

File: etna/models/holt_winters.py (hunk 3 above)
Comment (brsnw250) [reply etna/models/holt_winters.py]: Sure, it should be `min` to cover all cases

File: etna/models/utils.py [hunk 8]
```diff
@@ -98,3 +98,28 @@ def select_observations(
...
+    -------
+    :
+        pandas frequency string
+
+    Raises
+    ------
+    ValueError:
+        unable do determine frequency of data
+    """
+    freq = pd.infer_freq(timestamps, warn=False)
```
Comment (alex-hse-repository) [inline etna/models/utils.py:121]: Does it works the same in all the versions of pandas? In constructor of TSDataset we have try/except block for this case. May you add tests for this [here](https://github.com/tinkoff-ai/etna/blob/4b77a9941f5692a100081f236483c7ebe0bbee39/.github/workflows/test.yml#L221) -- to run the tests on different versions of pandas. May be it is also sensible to add all the tests form `datasets` forlder

File: etna/models/utils.py [hunk 9]
```diff
@@ -98,3 +98,28 @@ def select_observations(
     observations = observations.loc[timestamps]
     observations.reset_index(drop=drop_timestamp, inplace=True)
     return observations
+
+
+def determine_freq(timestamps: Union[pd.Series, pd.DatetimeIndex]) -> str:
+    """
+
```
Comment (alex-hse-repository) [inline etna/models/utils.py:105]: May you add description here

File: etna/models/utils.py [hunk 10]
```diff
@@ -98,3 +98,28 @@ def select_observations(
...
+    Parameters
+    ----------
+    timestamps
```
Comment (alex-hse-repository) [inline etna/models/utils.py:108]: There should be ":" in the end of the line

File: tests/test_models/test_utils.py [hunk 11]
```diff
@@ -104,3 +105,28 @@ def test_select_observations_without_timestamp(df_without_timestamp, timestamps)
...
+    )
+)
+def test_determine_freq(timestamps, answer):
+    assert determine_freq(timestamps=timestamps) == answer
+
+
+@pytest.mark.parametrize(
+    "timestamps",
+    (
+            pd.to_datetime(pd.Series(["2020-02-01", "2020-02-15", "2021-02-15"])),
```
Comment (alex-hse-repository) [inline tests/test_models/test_utils.py]: May you add case where there is only 1 and 2 points

File: etna/models/utils.py (hunk 5 above)
Comment (alex-hse-repository) [reply etna/models/utils.py:71]: In fact in all the cases we don't have timestamp column, may you simplify this method for our cases

File: tests/test_models/test_sarimax_model.py (hunk 7 above)
Comment (alex-hse-repository) [reply tests/test_models/test_sarimax_model.py:258]: Does it works properly without regressors? If so add test for it

File: etna/models/holt_winters.py [hunk 12]
```diff
@@ -372,9 +398,12 @@ def predict_components(self, df: pd.DataFrame) -> pd.DataFrame:
         model = self._model
         fit_result = self._result
 
-        if fit_result is None or model is None:
+        if fit_result is None or model is None or self._train_freq is None:
             raise ValueError("This model is not fitted!")
 
+        if df["timestamp"].min() < self._first_train_timestamp or df["timestamp"].max() > self._last_train_timestamp:
```
Comment (alex-hse-repository) [inline etna/models/holt_winters.py:404]: We actually only have test for the second case, but we don't check this one `df["timestamp"].min() < self._first_train_timestamp`

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/models/utils.py (hunk 5 above)
Comment (brsnw250) [reply etna/models/utils.py:71]: It is general purpose function that can be used in other places. Timeline could be provided within `df` or by function parameters.

File: tests/test_models/test_sarimax_model.py (hunk 7 above)
Comment (brsnw250) [reply tests/test_models/test_sarimax_model.py:258]: Actually without regressors we don't have any decomposition as we will return only SARIMA component.

File: tests/test_models/test_utils.py (hunk 11 above)
Comment (brsnw250) [reply tests/test_models/test_utils.py]: `pd.infer_freq` works with at least 3 points.

File: etna/models/utils.py (hunk 8 above)
Comment (brsnw250) [reply etna/models/utils.py:121]: If we look closely, we raise error if function fails to determine frequency. We can use try/except just to unify error message

File: etna/models/utils.py (hunk 9 above)
Comment (brsnw250) [reply etna/models/utils.py:105]: It is already there

File: etna/models/holt_winters.py (hunk 12 above)
Comment (brsnw250) [reply etna/models/holt_winters.py:404]: Added

File: etna/models/utils.py (hunk 5 above)
Comment (alex-hse-repository) [reply etna/models/utils.py:71]: In case we need this function somewhere we can enhance it, but now it just make it overcomplicated

**Review by alex-hse-repository — APPROVED** [review]

**Pull request merged**